src/common/components/markdownViewer/render/markdownRenderRules.ts

```diff
@@ -0,0 +1,89 @@
+import { MarkdownIt as Mdr } from 'markdown-it';
+import { SOURCE_LINE_CLASSNAME, SOURCE_LINE_ATTRIBUTE } from '../syncScroll';
+import { isHashHistory } from '../../../../history';
+
+// Factory to build custom rules easily.
+const CustomRuleFactory = (mdr: Mdr) => (name: string) => {
+  const originalRule = mdr.renderer.rules[name];
+  const customRule = {
+    ruleName: name,
+    rule: originalRule || mdr.renderer.renderToken.bind(mdr.renderer),
+    replace(newRule) {
+      this.rule = newRule;
+      return this;
+    },
+    decorate(decorator) {
+      const previousRule = this.rule;
+      this.rule = (tokens, idx, options, env, renderer) => {
+        decorator(tokens, idx, options, env, renderer);
+        return previousRule(tokens, idx, options, env, renderer);
+      };
+      return this;
+    },
+  };
+
+  return customRule;
+};
+
+// Decorator. It injects the source line number aimed for synch scroll.
+const sourceLineDecorator = (tokens, idx) => {
+  // Injection only for root-children tokens (level 0).
+  if (tokens[idx].map && tokens[idx].map.length && tokens[idx].level === 0) {
+    tokens[idx].attrJoin('class', SOURCE_LINE_CLASSNAME);
+    tokens[idx].attrSet(SOURCE_LINE_ATTRIBUTE, String(tokens[idx].map[0]));
+  }
+};
+
+// Load custom rules for a given markdown render.
+const loadCustomRules = (mdr: Mdr, routerLocation: string) => {
+  const CreateCustomRule = CustomRuleFactory(mdr);
+
+  // Add here desired customized rules.
+  // Keep source line number injection to all block elements for
+  // the synch scroll to work properly.
+  const customRules = [
+    CreateCustomRule('hr').decorate(sourceLineDecorator),
+    CreateCustomRule('heading_open').decorate(sourceLineDecorator),
+    CreateCustomRule('paragraph_open').decorate(sourceLineDecorator),
+    CreateCustomRule('image').decorate(sourceLineDecorator),
+    CreateCustomRule('code_block').decorate(sourceLineDecorator),
+    CreateCustomRule('fence').decorate(sourceLineDecorator),
+    CreateCustomRule('blockquote_open').decorate(sourceLineDecorator),
+    CreateCustomRule('list_item_open').decorate(sourceLineDecorator),
+    CreateCustomRule('ordered_list_open').decorate(sourceLineDecorator),
+    CreateCustomRule('bullet_list_open').decorate(sourceLineDecorator),
+    CreateCustomRule('table_open').decorate(sourceLineDecorator)
+      .decorate((tokens, idx) => {
```

Comment:
you can add typings like: `Token`

Reply:
Do you mean something like this? `tokens: Token[]`

Reply:
I guess strong typing, whenever we can enforce it is a good idea

Reply:
**Done**. Added strong types, now every parameter is typed directly or by inference.

Reply:
great